---
layout: post
title:  Testing Async functions in Mocha
date:    2017-06-04
excerpt: Learn You An Async For Great Good
image: /assets/img/covers/mocha.jpg
tag:
  - Javascript
  - Node
  - Testing
---

Do you know what day it is? One month until Independence Day? Nope, it's nothing actually, I just wanted to check that you weren't forgetting anything.

In other news, I finally tried to write some tests for [pkgparse](https://github.com/marcus-crane/pkgparse), the module that I use to test out new Node features or styles of writing things.

I changed it to use async/await instead of promises to celebrate the release of Node [v8.0.0](https://nodejs.org/en/blog/release/v8.0.0/).

Now, you might say that async/await has been in Node since 7.7.4 which, yes that's true but 8.X.X is the next Long Term Support version!

The team haven't said explicitly that it's 8.0.0 itself but I imagine we'll start to see it looked at more until, and especially after the LTS version is announced

I haven't put in 1000 hours of testing time yet so writing tests is still something I'm learning a lot about but testing async functions is definitely a new one for me. That is to say, I've tested promises but in order to await the function you're testing, the actual test function needs to use await.

Well, just use `done()` I hear you say but the trouble I had was that using `expect()` followed by `done()` is mixing promises and callbacks which is a no, no according to mocha. There were a few comments around suggesting that so I don't know what the problem was.

Here's what I've since learnt anyway. We'll use the following function as a test example:

```js
module.exports = async function fetchDependenciesFromNPM (moduleName) {
  try {
    const JSONObjectFromNPM = await axios('https://registry.npmjs.org/' + moduleName)
    // some filtering here blah blah
    if (dependenciesExist) {
      return dependencies // eg; ['mocha', 'express', 'axios']
    } else {
      return []
    }
  } catch (error) {
    throw new Error(`${error.response.status} ${error.response.statusText}`)
  }  
}
```

Most of this we can ignore for this example. The only things we need to know is that this function takes the name of a module as a string eg: `express` and returns its NPM dependencies as an array.

If there's not dependencies at all, it just sends back an empty array instead. In reality, pkgparse uses a Set but I'm still figuring out how to deal with the contents of a Set that has no keys. Testing its size/length is easy enough.

Anyway, I'm getting sidetracked here. Here's how I'm currently testing the above function:

```js
it('should return the dependencies of a given NPM module', async function() {
  const dependencies = await fetchDependenciesFromNPM('yourfavemodule')
  return expect(dependencies).to.deep.equal([ 'eslint', 'react', '' ])
})
```

I was fiddling with `try/catch` blocks at first and `done()` but apparently we can just use the regular `expect()` callback which makes sense seeing as you can almost treat `await`-ed functions like a regular synchronous functions.

This raises the question about how to test thrown errors however. I fiddled around a bit and using the above gives you no real way to do that. One possible solution doesn't require too much extra code though:

```js 
it('should throw an error when given an invalid module name', async function() {
  try {
    await fetchDependenciesFromNPM('blahblah')
  } catch (error) {
    expect(error.message).to.equal('404 Not Found')
  }
})
```

I'm not exactly sure if it's best practice but it worksâ„¢. I'll let you know if I have a realisation in the future that it's all wrong!